### PR TITLE
Init optimizer

### PR DESCRIPTION
**Pull request by RodionovF**

### CODE REVIEW

**Comment by RodionovF** [comment]
add tests

File: src/optimizer.cpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
+#include <deepworks/optimizer.hpp>
+
+namespace deepworks {
+namespace optimizer {
+
+using ConstMatrix = Eigen::Map<const Eigen::Matrix<float, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>>;
+using Matrix = Eigen::Map <Eigen::Matrix<float, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>>;
+
+SGD::SGD(int& params, float lr) : parameters(params), learning_rate(lr) {}
```
Comment (TolyaTalamanov) [inline src/optimizer.cpp:9]: Hod does it work ? You pass `int&` as `Parameters&`

File: src/optimizer.cpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+void SGD::step() {
+    for (auto& parameter: parameters) {
+        if (parameter.is_trainable()) {
+            auto weight = parameter.data();
+            const auto grad = parameter.grad();
+
+            const auto& shape = weight.shape();
+            DeepWorks_Assert(shape.size() == 2);
```
Comment (TolyaTalamanov) [inline src/optimizer.cpp:18]: What about bias ? I wouldn't check such kind of logic in there

File: src/optimizer.cpp [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+float SGD::get_learning_rate() {
+    return learning_rate;
+}
+
+float SGD::set_learning_rate(float lr) {
+    learning_rate = lr;
+}
+
+} // namespace optimizer
+} // namespace deepworks
```
Comment (TolyaTalamanov) [inline src/optimizer.cpp:38]: New line

File: include/deepworks/optimizer.hpp [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    explicit SGD(Parameters& params, float lr);
+
+    void step();
+
+    float get_learning_rate();
+
+    void set_learning_rate(float lr);
+
+private:
+    float learning_rate;
```
Comment (TolyaTalamanov) [inline include/deepworks/optimizer.hpp:21]: to keep consistency I would name it `m_lr` or `m_learning_rate` and `m_parameters` or `m_params`

File: include/deepworks/optimizer.hpp [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+#include <deepworks/tensor.hpp>
+#include <deepworks/parameter.hpp>
+
+namespace deepworks {
+namespace optimizer {
+
+class SGD {
+public:
+
```
Comment (TolyaTalamanov) [inline include/deepworks/optimizer.hpp:11]: Extra line

**Comment by l-bat** [comment]
Please don't forget to include `include/deepworks/optimizer.hpp` to `include/deepworks/deepworks.hpp`

**Comment by RodionovF** [comment]
@KSTconsult @TolyaTalamanov @apertovs @l-bat

File: src/optimizer.cpp [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+void SGD::step() {
+    for (auto& param: m_params) {
+        if (param.is_trainable()) {
+            auto weight = param.data();
+            const auto grad = param.grad();
+
+            const auto& shape = weight.shape();
+
+            DeepWorks_Assert(grad.shape() == shape);
+            DeepWorks_Assert(shape.size() == 2 || shape.size() == 1);
```
Comment (l-bat) [inline src/optimizer.cpp:23]: For convolution `shape.size() = 4`. I think we don't need this check.

File: tests/unit/test_optimizer.cpp [hunk 7]
```diff
@@ -0,0 +1,82 @@
+#include <gtest/gtest.h>
+
+//#include <cmath>
+
+//#include "kernels_reference.hpp"
+//#include "test_utils.hpp"
+//#include <deepworks/initializers.hpp>
```
Comment (l-bat) [inline tests/unit/test_optimizer.cpp:7]: remove?

File: src/optimizer.cpp (hunk 6 above)
Comment (RodionovF) [reply src/optimizer.cpp]: ty, fixed it
i live in 2d world

**Comment by TolyaTalamanov** [comment]
Thanks! LGTM :+1:

**Pull request merged**